drivers: intc: plic: implement `riscv_plic_irq_set_pending()`

ycsin opened the pull request:
Set an interrupt line as pending for PLIC that supports software-generated interrupt. Currently, this is only supported by Andes NCEPLIC100 (`"andestech,nceplic100"`), other PLICs can be supported by updating the Kconfig.

For more info, see https://github.com/zephyrproject-rtos/zephyr/pull/78611#issuecomment-2371082864

For context, this can be used for SMP IPI / mailbox

ycsin committed:
drivers: intc: plic: implement software-generated interrupt

Implement `riscv_plic_irq_set_pending()` to trigger a
software-generated interrupt.

The "4. Interrupt Pending Bits" of the riscv-plic specs
described the reading of the pending bits, but not the writing

Since not all PLIC implementations support software-generated
interrupt, the function is compiled only when
`CONFIG_PLIC_SUPPORTS_SOFT_INTERRUPT` is enabled on PLIC that
supports it, such as the Andes' NCEPLIC100.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>
Signed-off-by: Yong Cong Sin <yongcong.sin@gmail.com>

jimmyzhe commented:
Thank you for this PR, Andes platform works with this change. I plan to rework the mbox_andes_plic_sw driver after this PR is merged.

But I think not all PLIC implementation support software-triggered interrupt. 
The RISC-V PLIC spec defines the Pending Bit can be read and cleared by the claim operation ,but it doesn't describe write operation. Some PLIC implementation mark Pend Bit as read-only.
- [FE310-G000 Manual v3p2](https://www.starfivetech.com/uploads/fe310-g000-manual-v3p2.pdf) <img width="400" alt="擷取60" src="https://github.com/user-attachments/assets/d0573e5f-563a-4242-9a38-0dfa3fd8b175">

Andes PLIC describes it as an enhanced feature
- [AndeStar V5 PLIC Specification](https://www.andestech.com/en/products-solutions/product-documentation/) (it needs to fill some personal info to download the document, sorry I can't share it diretly.) <img width="438" alt="擷取57" src="https://github.com/user-attachments/assets/dd365e42-5778-4b63-a364-f3f37d0d533f">

Maybe use `#if IS_ENABLED(PLIC_DRV_HAS_COMPAT(andestech_nceplic100))` or add a Kconfig for PLICs that do not support this feature would be better.

ycsin committed:
dts: add andestech,nceplic100 binding

Add a new `andestech,nceplic100` binding that inherits from the
`sifive,plic-1.0.0` binding. This is so that the Kconfig
`DT_HAS_ANDESTECH_NCEPLIC100_ENABLED` would be generated during
build.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>
Signed-off-by: Yong Cong Sin <yongcong.sin@gmail.com>

ycsin committed:
drivers: intc: plic: convert trigger type to use Kconfig

Convert the compilation of the trigger type feature to depend
on Kconfig, following the same pattern of software-triggered
interrupt.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>
Signed-off-by: Yong Cong Sin <yongcong.sin@gmail.com>

ycsin commented:
@jimmyzhe thanks for pointing that out. I've updated the patch accordingly.

> I plan to rework the mbox_andes_plic_sw driver after this PR is merged.

Could you probably comment your plan in #78917, so that Andes mbox works alongside SMP? I can be reached with the same handle on Discord as well.

ycsin commented on drivers/interrupt_controller/Kconfig.plic:
```diff
@@ -13,6 +13,44 @@ config PLIC
+	help
+	  Offset to the 'trigger type' register.
+
+config PLIC_TRIG_TYPE_WIDTH
+	int "Trigger type width"
```
"bitwidth" is probably a better term

```suggestion
config PLIC_TRIG_TYPE_BITWIDTH
	int "Trigger type bitwidth"
```

ycsin commented on drivers/interrupt_controller/Kconfig.plic:
```diff
@@ -13,6 +13,44 @@ config PLIC
+config PLIC_TRIG_TYPE_WIDTH
+	int "Trigger type width"
+	default 1 if DT_HAS_ANDESTECH_NCEPLIC100_ENABLED
+	help
```
I purposely left this blank so that the driver would fail to compile if some OOT configuration selected `PLIC_SUPPORTS_TRIG_TYPE` but forgot to configure these (same as `PLIC_TRIG_TYPE_REG_OFFSET`).

Is a default value necessary? If trigger type is supported, this should be >= 1

cfriedt replied:
Maybe make the default 1 then, independent of which plic implementation it is.

If other plic implementations have larger bit widths, then there should be default specialization

ycsin commented:
🙌
